edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs

```diff
@@ -27,23 +31,54 @@ public Task StoreMetricsAsync(IEnumerable<Metric> metrics)
 
         public async Task<IEnumerable<Metric>> GetAllMetricsAsync()
         {
-            return (await this.dataStore.GetBatch(0, 1000))
-                    .SelectMany(((long offset, IEnumerable<Metric> metrics) storeResult) =>
+            // Note: async/await is needed b/c ToListAsync returns ValueTask instead of Task.
+            return await this.GetAllMetricsInternal().ToListAsync();
+        }
+
+        async IAsyncEnumerable<Metric> GetAllMetricsInternal([EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            HashSet<long> maxAgeEntriesToRemove = new HashSet<long>();
+
+            foreach ((long offset, IEnumerable<Metric> metrics) in await this.dataStore.GetBatch(0, 1000, cancellationToken))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                // This makes the assumption that metrics are stored in batches with matching timestamps.
+                // This is the case b/c the metrics are stored immediately after being scraped.
+                if (metrics.First().TimeGeneratedUtc < DateTime.UtcNow - this.maxMetricAge)
```

Comment:
`UtcNow` will be reevaluated each time through the loop, so different batches will be evaluated against different max-age timestamps. Is that the right behavior or do we want to take a snapshot of `UtcNow` before we enter the loop?

Reply:
It should happen so fast it won't matter (we're comparing on a week timescale), but I'll go ahead and change it for consistency